client_test.go

```diff
@@ -612,3 +613,99 @@ func TestNewWithLocalAddr(t *testing.T) {
 	assertEqual(t, resp.String(), "TestGet: text response")
 }
 
+func TestClientOnResponseError(t *testing.T) {
+	ts := createAuthServer(t)
+	defer ts.Close()
+
+	tests := []struct {
+		name    string
+		setup   func(*Client)
+		isError bool
+	}{
+		{
+			name: "successful_request",
+		},
+		{
+			name: "http_status_error",
+			setup: func(client *Client) {
+				client.SetAuthToken("BAD")
+			},
+		},
+		{
+			name: "before_request_error",
+			setup: func(client *Client) {
+				client.OnBeforeRequest(func(client *Client, request *Request) error {
+					return fmt.Errorf("before request")
+				})
+			},
+			isError: true,
+		},
+		{
+			name: "before_request_error_retry",
+			setup: func(client *Client) {
+				client.SetRetryCount(3).OnBeforeRequest(func(client *Client, request *Request) error {
+					return fmt.Errorf("before request")
+				})
+			},
+			isError: true,
+		},
+		{
+			name: "after_response_error",
+			setup: func(client *Client) {
+				client.OnAfterResponse(func(client *Client, response *Response) error {
+					return fmt.Errorf("after response")
+				})
+			},
+			isError: true,
+		},
+		{
+			name: "after_response_error_retry",
+			setup: func(client *Client) {
+				client.SetRetryCount(3).OnAfterResponse(func(client *Client, response *Response) error {
+					return fmt.Errorf("after response")
+				})
+			},
+			isError: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			var errs int
+			c := New().outputLogTo(ioutil.Discard).
+				SetTLSClientConfig(&tls.Config{InsecureSkipVerify: true}).
+				SetAuthToken("004DDB79-6801-4587-B976-F093E6AC44FF").
+				SetRetryCount(0).
+				SetRetryMaxWaitTime(time.Microsecond).
+				AddRetryCondition(func(response *Response, err error) bool {
+					if err != nil {
+						return true
+					}
+					return response.IsError()
+				}).
+				OnError(func(r *Request, err error) {
+					errs++
```

Comment:
Should access `r` in this test to ensure that we're properly passing the request into the hook.

Also, it doesn't look like you're testing multiple OnError hooks in a single request.  If we're going to keep that ability, we ought to test it.

Reply:
> One question: why allow multiple `OnError` hooks to be defined? What's the use case?

Don't have a particular use-case in mind for multiple on-error hooks, but it seemed like using a a range makes it at least possible for one to arise without a code change to accommodate it; and it doesn't make the implementation any less clean.